### ISSUE TITLE
Error with Tail Recursion when using Console.readLine()

### ISSUE THREAD

**Issue by billpcs** [flix/flix#754]
Hello!

I recently discovered **Flix** and I am writing some small programs! I really enjoy the language.

Given that the standard library offers `readLine()` I wanted to create `readLines()`. It is _not_ a great implementation as it checks for empty line, not EOF, but anyway. Here is what I wrote.

```scala
def readLinesRec(xs: List[Str]): List[Str] = {
  let line = Console.readLine();
  match line with {
    case "" => List.reverse(xs)
    case _  => readLinesRec(line :: xs)
  }
}
```

I then invoke the function from main with

```scala
readLinesRec(Nil)
```
When I try to do `java -jar flix.jar run --verbose` on the project, I get the following 
[stack trace](https://github.com/flix/flix/files/2748893/out.txt). I am running v0.4.0.

I know this function works when not invoking `readLine`, if for example it's consuming a list instead.

So I was wondering, what [am I doing/is going] wrong?

Cheers!

**Comment by magnus-madsen** [comment]
Hello!

Thanks for trying out Flix! There is nothing wrong with your program :) 

You have found a bug in the compiler! The fix is not easy, so it will take a while to implement properly. 

[Meanwhile I have released 0.4.1](https://github.com/flix/flix/releases) which includes a new flag `--Xno-stratifier` which disables the faulty phase of the compiler. (This phase is only needed for first-class constraints, so if you don't use those, it is safe to disable).

You can now run your program:

```scala
C:\Work\GitHub\flix>java -jar flix.jar foo.flix --Xno-stratifier
abc
hello world!

Cons((abc, Cons((hello world!, Nil(())))))
```

**Comment by billpcs** [comment]
Amazing! Yes, indeed, it now works.

So is this bug related to `readLine` per se or is it something more general?

**Comment by magnus-madsen** [comment]
It's unrelated to `readLine`. It's a bug in the control-flow analysis which is used for the logic part of Flix.

**Comment by magnus-madsen** [comment]
This bug has been fixed in master. It will be available in the next release of Flix (0.7) scheduled for sometime in the summer.
